cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i

```diff
@@ -677,6 +677,7 @@ services::Status TrainBatchTaskBase<algorithmFPType, BinIndexType, DataHelper, c
     _nConstFeature           = 0;
     _numElems                = &numElems;
     _engineImpl              = engineImpl;
+    _helper._engineImpl      = engineImpl;
```

Comment:
It seems now there are 2 copies of engineImpl are stored in TrainBatchTaskBase:

- TrainBatchTaskBase::_engineImpl
- TrainBatchTaskBase::_helper._engineImpl

This can lead to an inconsistent state of TrainBatchTaskBase object. I think it would be better to remove TrainBatchTaskBase::_engineImpl field at all and use TrainBatchTaskBase::_helper._engineImpl instead.
It would look like just moving a pointer to random number generator from Task into the Helper object.

Reply:
Completed in latest commits, will have to retest code to make sure results stay similar (but shouldn't be a problem)